### PR TITLE
Error when non-existent file is encountered 

### PR DESCRIPTION
**Pull request by AnthonyBroadCrawford**
# PR for issue #3037 
This PR is now development complete and ready for review. 

## Decisions 
When implementing this feature I took the following approach 

1. Error and exit compilation if a non-existent file is encountered 
2. Create a new flag --ignore-missing that will continue compilation rather than error out when a missing file is encountered

## Questions 
Given I'm still fairly new to the codebase & Solidity in general I have the following questions 

1. In the method where I discover a missing file *and* the user hasn't passed in the `--ignore-missing` option I leverage `exit(1)` as a means to escape.  Primarily, it was the easiest path vs having to pass a boolean up the stack to inspect and then return 1/false further up the stack. 

2.  Please jump on my copy.  I don't have a feel for the prose throughout the code base for error message/communications within the compiler.  Open to any/all feedback. 

## Sample commands 
Here's a simple way to test it when compiling this branch.  This is executed from the ***build*** directory with the attached simple contract located in the same ***build*** directory 

*NO ERROR: * `./solc/solc --bin --ignore-missing test.sol missing-file.sol`
*ERROR* `./solc/solc --bin test.sol missing-file.sol`

## Sample contract 
Again, my above commands assume this is located in the *build* director 

```
pragma solidity ^0.4.17;

contract Test {

}
```

### CODE REVIEW

File: solc/CommandLineInterface.cpp [hunk 1]
```diff
@@ -416,8 +418,13 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
 				auto infile = boost::filesystem::path(path);
 				if (!boost::filesystem::exists(infile))
 				{
-					cerr << "Skipping non-existent input file \"" << infile << "\"" << endl;
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Just turn this into an error. It should not be an option that you can switch on or off.

File: solc/CommandLineInterface.cpp [hunk 2]
```diff
@@ -484,23 +491,33 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
 	return true;
 }
 
+
+//
+//HERE - Error out instead of skipping non-existent source files #3037
+//
 void CommandLineInterface::createFile(string const& _fileName, string const& _data)
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: This is the routine that writes output files, you should not need to touch this.

Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @chriseth Yeah, I plan on removing that comment as it was a placeholder for me to simply orient myself in the code.

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @chriseth we want to silently ignore the missing file?  Is there a reason we wouldn't to simply inform the user we are skipping it (for education, debugging, general UX)?  Not taking stance here, simply seeking to better understand your POV.  Thanks for all the feedback so far.

**Comment by AnthonyBroadCrawford** [comment]
@chriseth I took your feedback and only error when a file is missing and command line flag to suppress is not provided.  Updated change log and squashed down to a single commit.  Let me know if you see anything else that needs changed.

**Comment by AnthonyBroadCrawford** [comment]
Looking into the failing tests.

**Comment by axic** [comment]
Please use rebase and not merges.

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -416,8 +418,10 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
...
 				{
-					cerr << "Skipping non-existent input file \"" << infile << "\"" << endl;
-					continue;
+					if(! m_args.count(g_argIgnoreMissingFiles)){
+						cerr << "Non-existent input file " << infile << endl;
+						exit(1);
+					}
 				}
 
 				if (!boost::filesystem::is_regular_file(infile))
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: I think this should be considered part of this. It just does skipping if "file is invalid".

Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @axic Ah, I can update that section as well.  To be clear, if the file is missing OR invalid I should communicate to the user and error out.  That's what you are requesting.

Comment (axic) [reply solc/CommandLineInterface.cpp]: @chriseth are you also ok with that?

File: Changelog.md [hunk 4]
```diff
@@ -27,6 +27,9 @@ Features:
  * Syntax Checker: Issue error if no visibility is specified on contract functions as experimental 0.5.0 feature.
  * Syntax Checker: Issue warning when using overloads of ``address`` on contract instances.
  * Type Checker: disallow combining hex numbers and unit denominations as experimental 0.5.0 feature.
+ * Improved messaging when error spans multiple lines of a sourcefile 
+ * Commandline interface:  Error when missing file detected. Support to suppress with missing error ``--ignore-missing`` flag.
```
Comment (chriseth) [inline Changelog.md]: Please move this to the 0.4.22 section.

File: Changelog.md [hunk 5]
```diff
@@ -27,6 +27,9 @@ Features:
  * Syntax Checker: Issue error if no visibility is specified on contract functions as experimental 0.5.0 feature.
  * Syntax Checker: Issue warning when using overloads of ``address`` on contract instances.
  * Type Checker: disallow combining hex numbers and unit denominations as experimental 0.5.0 feature.
+ * Improved messaging when error spans multiple lines of a sourcefile
```
Comment (chriseth) [inline Changelog.md]: This should not be part of this PR.

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (chriseth) [reply solc/CommandLineInterface.cpp]: No, just stop processing. I'm not sure we need an option to allow files to be missing.

File: solc/CommandLineInterface.cpp [hunk 6]
```diff
@@ -416,8 +418,10 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
 				auto infile = boost::filesystem::path(path);
 				if (!boost::filesystem::exists(infile))
 				{
-					cerr << "Skipping non-existent input file \"" << infile << "\"" << endl;
-					continue;
+					if(! m_args.count(g_argIgnoreMissingFiles)){
+						cerr << "Non-existent input file " << infile << endl;
+						exit(1);
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Do other parts of the code directly call exit? If not, rather return a bool.

File: solc/CommandLineInterface.cpp (hunk 3 above)
Comment (chriseth) [reply solc/CommandLineInterface.cpp]: yes

File: solc/CommandLineInterface.cpp [hunk 7]
```diff
@@ -487,20 +491,26 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
 void CommandLineInterface::createFile(string const& _fileName, string const& _data)
 {
 	namespace fs = boost::filesystem;
+
 	// create directory if not existent
-	fs::path p(m_args.at(g_argOutputDir).as<string>());
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Please do not touch this function in this PR.

**Review by chriseth — CHANGES_REQUESTED** [review]
@axic what is your opinion about having the switch?

File: Changelog.md (hunk 4 above)
Comment (AnthonyBroadCrawford) [reply Changelog.md]: @chriseth Yup, my apologies.  I'll have an updated and squashed (removing merge) tonight.

File: solc/CommandLineInterface.cpp (hunk 3 above)
Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @axic @chriseth Alright, officially added as scope.

File: solc/CommandLineInterface.cpp (hunk 7 above)
Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @chriseth My apologies, I didn't mean to have those make it into the PR.  I will remove any touch to this function.

File: solc/CommandLineInterface.cpp (hunk 6 above)
Comment (AnthonyBroadCrawford) [reply solc/CommandLineInterface.cpp]: @chriseth  I couldn't find other instances so I can return a bool.  It's gonna cause a little more work further up the stack which is not a big deal, simply expands the scope for number of functions this change touches.  I'll have it return a bool.

**Comment by chriseth** [comment]
@AnthonyBroadCrawford this is still not properly rebased. Please tell us if we should do it for you.
